network/retrieve/retrieve.go

```diff
@@ -0,0 +1,432 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package retrieve
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/network/timeouts"
+	"github.com/ethersphere/swarm/p2p/protocols"
+	"github.com/ethersphere/swarm/spancontext"
+	"github.com/ethersphere/swarm/storage"
+	opentracing "github.com/opentracing/opentracing-go"
+	olog "github.com/opentracing/opentracing-go/log"
+)
+
+var (
+	// Compile time interface check
+	_ node.Service = &Retrieval{}
+
+	// Metrics
+	processReceivedChunksCount    = metrics.NewRegisteredCounter("network.retrieve.received_chunks.count", nil)
+	handleRetrieveRequestMsgCount = metrics.NewRegisteredCounter("network.retrieve.handle_retrieve_request_msg.count", nil)
+	retrieveChunkFail             = metrics.NewRegisteredCounter("network.retrieve.retrieve_chunks_fail.count", nil)
+
+	lastReceivedRetrieveChunksMsg = metrics.GetOrRegisterGauge("network.retrieve.received_chunks", nil)
+
+	// Protocol spec
+	spec = &protocols.Spec{
+		Name:       "bzz-retrieve",
+		Version:    1,
+		MaxMsgSize: 10 * 1024 * 1024,
+		Messages: []interface{}{
+			ChunkDelivery{},
+			RetrieveRequest{},
+		},
+	}
+
+	ErrNoPeerFound = errors.New("no peer found")
+)
+
+// Retrieval holds state and handles protocol messages for the `bzz-retrieve` protocol
+type Retrieval struct {
+	mtx      sync.Mutex
+	netStore *storage.NetStore
+	kad      *network.Kademlia
+	peers    map[enode.ID]*Peer //this protocols peers
+	spec     *protocols.Spec    //this protocol's spec
+	balance  protocols.Balance  //implements protocols.Balance, for accounting
+	prices   protocols.Prices   //implements protocols.Prices, provides prices to accounting
+
+	quit chan struct{} // termination
+}
+
+// NewRetrieval returns a new instance of the retrieval protocol handler
+func NewRetrieval(kad *network.Kademlia, ns *storage.NetStore) *Retrieval {
+	ret := &Retrieval{
+		kad:      kad,
+		peers:    make(map[enode.ID]*Peer),
+		netStore: ns,
+		quit:     make(chan struct{}),
+	}
+
+	ret.spec = spec
+
+	return ret
+}
+
+func (r *Retrieval) addPeer(p *Peer) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.peers[p.ID()] = p
+}
+
+func (r *Retrieval) removePeer(p *Peer) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	delete(r.peers, p.ID())
+}
+
+func (r *Retrieval) getPeer(id enode.ID) *Peer {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	return r.peers[id]
```

Comment:
What would happen if we are concurrently handling 2 messages from the same peer, one of them resulting in an error, which would return from the `Run` method below, and call `removePeer`, while the other is still processing a message and then calling `getPeer` and `send` on a nil pointer?

I think it is a bit dangerous to assume that a peer is always in this slice, this is why I have added checks in the `PSS` PR.

I guess you can leave it as is, but keep it at the back of your head in case we ever see nil pointer dereferences on `Send`.

Reply:
Related / similar: https://github.com/ethersphere/swarm/pull/1580/files#diff-6109308aa1312b43e9a8d048e2f2d1ffR284